File: tools/agentstream/tiedpreflight/tiedpreflight.go
// Package tiedpreflight validates .cursor/mcp.json for tied-yaml before spawning cursor agent.
// REQ-GOAGENT-CLI-CONFIG (companion guard)
package tiedpreflight

import (
	"encoding/json"
	"errors"
	"fmt"
	"os"
	"path/filepath"
	"strings"
)

// ErrNotFound is returned when no .cursor/mcp.json is discovered.
var ErrNotFound = errors.New("tiedpreflight: no .cursor/mcp.json found under workspace")

// ErrAmbiguous is returned when multiple subproject mcp.json files exist without an explicit path.
type ErrAmbiguous struct {
	Paths []string
}

func (e *ErrAmbiguous) Error() string {
	return fmt.Sprintf("tiedpreflight: multiple .cursor/mcp.json files under workspace; pass --mcp-json PATH:\n  %s", strings.Join(e.Paths, "\n  "))
}

// Result summarizes MCP config checks (not runtime Cursor MCP registration).
type Result struct {
	MCPJSONPath   string
	TIEDBasePath  string
	Status        Status
	Errors        []string
	Warnings      []string
	RuntimeNotice string
}

// Status is the worst severity from static checks.
type Status int

const (
	// StatusOK means tied-yaml entry and TIED_BASE_PATH look sane for this workspace.
	StatusOK Status = iota
	// StatusWarning means proceed with caution (e.g. greenfield tied/, or informational only).
	StatusWarning
	// StatusBlocked means configuration is missing or dangerous (e.g. TIED_BASE_PATH outside workspace).
	StatusBlocked
)

// RuntimeMCPNotice is printed whenever preflight runs to remind that Cursor may still omit tied-yaml.
const RuntimeMCPNotice = "agentstream cannot verify that Cursor exposes the tied-yaml MCP server for this workspace; enable tied-yaml in Cursor if agents should edit project TIED YAML via MCP."

// LocateMCPJSON finds .cursor/mcp.json: explicit override, workspace root, or exactly one immediate subdir.
func LocateMCPJSON(workspace, override string) (string, error) {
	workspace = filepath.Clean(workspace)
	if override != "" {
		p := filepath.Clean(override)
		st, err := os.Stat(p)
		if err != nil {
			return "", fmt.Errorf("tiedpreflight: --mcp-json not readable: %w", err)
		}
		if st.IsDir() {
			return "", fmt.Errorf("tiedpreflight: --mcp-json is a directory, expected a file: %s", p)
		}
		return p, nil
	}
	root := filepath.Join(workspace, ".cursor", "mcp.json")
	if st, err := os.Stat(root); err == nil && !st.IsDir() {
		return root, nil
	}
	matches, err := filepath.Glob(filepath.Join(workspace, "*", ".cursor", "mcp.json"))
	if err != nil {
		return "", err
	}
	switch len(matches) {
	case 0:
		return "", ErrNotFound
	case 1:
		return matches[0], nil
	default:
		return "", &ErrAmbiguous{Paths: matches}
	}
}

type mcpFile struct {
	MCPServers map[string]json.RawMessage `json:"mcpServers"`
}

type tiedYAMLServer struct {
	Env  map[string]string `json:"env"`
	Args []string          `json:"args"`
}

// Analyze reads mcp.json and validates tied-yaml / TIED_BASE_PATH against workspace.
func Analyze(workspace, mcpJSONPath string) (*Result, error) {
	workspace = filepath.Clean(workspace)
	res := &Result{
		MCPJSONPath:   mcpJSONPath,
		RuntimeNotice: RuntimeMCPNotice,
	}
	data, err := os.ReadFile(mcpJSONPath)
	if err != nil {
		res.Status = StatusBlocked
		res.Errors = append(res.Errors, fmt.Sprintf("read mcp.json: %v", err))
		return res, nil
	}
	var root mcpFile
	if err := json.Unmarshal(data, &root); err != nil {
		res.Status = StatusBlocked
		res.Errors = append(res.Errors, fmt.Sprintf("parse mcp.json: %v", err))
		return res, nil
	}
	if root.MCPServers == nil {
		res.Status = StatusBlocked
		res.Errors = append(res.Errors, "mcp.json has no mcpServers object")
		return res, nil
	}
	raw, ok := root.MCPServers["tied-yaml"]
	if !ok || len(raw) == 0 {
		res.Status = StatusBlocked
		res.Errors = append(res.Errors, `mcpServers["tied-yaml"] is missing; run copy_files.sh on this project or add tied-yaml in .cursor/mcp.json`)
		return res, nil
	}
	var srv tiedYAMLServer
	if err := json.Unmarshal(raw, &srv); err != nil {
		res.Status = StatusBlocked
		res.Errors = append(res.Errors, fmt.Sprintf("parse tied-yaml server entry: %v", err))
		return res, nil
	}
	base := strings.TrimSpace(srv.Env["TIED_BASE_PATH"])
	if base == "" {
		res.Status = StatusBlocked
		res.Errors = append(res.Errors, `tied-yaml env.TIED_BASE_PATH is missing or empty`)
		return res, nil
	}
	res.TIEDBasePath = base
	if !filepath.IsAbs(base) {
		res.Status = StatusBlocked
		res.Errors = append(res.Errors, fmt.Sprintf("TIED_BASE_PATH must be absolute (got %q); see TIED docs for .cursor/mcp.json", base))
		return res, nil
	}
	wAbs, err := filepath.EvalSymlinks(workspace)
	if err != nil {
		wAbs = workspace
	}
	tAbs, err := filepath.EvalSymlinks(filepath.Clean(base))
	if err != nil {
		res.Status = StatusBlocked
		res.Errors = append(res.Errors, fmt.Sprintf("TIED_BASE_PATH does not resolve: %v", err))
		return res, nil
	}
	rel, err := filepath.Rel(wAbs, tAbs)
	if err != nil || rel == ".." || strings.HasPrefix(rel, ".."+string(filepath.Separator)) {
		res.Status = StatusBlocked
		res.Errors = append(res.Errors, fmt.Sprintf("TIED_BASE_PATH (%s) is outside --workspace (%s); MCP writes may target the wrong repo", tAbs, wAbs))
		return res, nil
	}
	req := filepath.Join(tAbs, "requirements.yaml")
	if st, err := os.Stat(req); err != nil || st.IsDir() {
		res.Warnings = append(res.Warnings, "no readable tied/requirements.yaml at TIED_BASE_PATH (greenfield or partial bootstrap)")
		if res.Status == StatusOK {
			res.Status = StatusWarning
		}
	}
	if res.Status == StatusOK && len(res.Warnings) > 0 {
		res.Status = StatusWarning
	}
	return res, nil
}

// Run locates and analyzes MCP config. If override is empty, searches under workspace.
func Run(workspace, mcpJSONOverride string) (*Result, error) {
	path, err := LocateMCPJSON(workspace, mcpJSONOverride)
	if err != nil {
		return nil, err
	}
	return Analyze(workspace, path)
}

File: mcp-server/packages/agentstream/src/tiedpreflight.ts
/**
 * [IMPL-TIED_UNIFIED_TOOLCHAIN] [ARCH-TIED_UNIFIED_TOOLCHAIN] [REQ-GOAGENT-CLI-CONFIG]
 * Static tied-yaml MCP preflight — TS port of tools/agentstream/tiedpreflight.
 */
import fs from "node:fs";
import path from "node:path";

export const ERR_NOT_FOUND = "tiedpreflight: no .cursor/mcp.json found under workspace";

export class ErrAmbiguous extends Error {
  readonly paths: string[];

  constructor(paths: string[]) {
    super(
      `tiedpreflight: multiple .cursor/mcp.json files under workspace; pass --mcp-json PATH:\n  ${paths.join("\n  ")}`,
    );
    this.name = "ErrAmbiguous";
    this.paths = paths;
  }
}

export const Status = {
  OK: 0,
  Warning: 1,
  Blocked: 2,
} as const;

export type Status = (typeof Status)[keyof typeof Status];

export const RuntimeMCPNotice =
  "agentstream cannot verify that Cursor exposes the tied-yaml MCP server for this workspace; enable tied-yaml in Cursor if agents should edit project TIED YAML via MCP.";

export type PreflightResult = {
  mcpJsonPath: string;
  tiedBasePath: string;
  status: Status;
  errors: string[];
  warnings: string[];
  runtimeNotice: string;
};

type McpFile = {
  mcpServers?: Record<string, unknown>;
};

type TiedYamlServer = {
  env?: Record<string, string>;
  args?: string[];
};

export function locateMcpJson(workspace: string, override: string): string {
  const ws = path.resolve(workspace);
  if (override.trim() !== "") {
    const p = path.resolve(override);
    let st: fs.Stats;
    try {
      st = fs.statSync(p);
    } catch (err) {
      throw new Error(`tiedpreflight: --mcp-json not readable: ${String(err)}`);
    }
    if (st.isDirectory()) {
      throw new Error(
        `tiedpreflight: --mcp-json is a directory, expected a file: ${p}`,
      );
    }
    return p;
  }

  const root = path.join(ws, ".cursor", "mcp.json");
  try {
    const st = fs.statSync(root);
    if (!st.isDirectory()) {
      return root;
    }
  } catch {
    // fall through
  }

  let entries: fs.Dirent[];
  try {
    entries = fs.readdirSync(ws, { withFileTypes: true });
  } catch (err) {
    throw err;
  }

  const matches: string[] = [];
  for (const ent of entries) {
    if (!ent.isDirectory()) {
      continue;
    }
    const candidate = path.join(ws, ent.name, ".cursor", "mcp.json");
    try {
      const st = fs.statSync(candidate);
      if (!st.isDirectory()) {
        matches.push(candidate);
      }
    } catch {
      // skip
    }
  }

  if (matches.length === 0) {
    throw new Error(ERR_NOT_FOUND);
  }
  if (matches.length === 1) {
    return matches[0]!;
  }
  throw new ErrAmbiguous(matches.sort());
}

function resolveSymlinkSafe(p: string): string {
  try {
    return fs.realpathSync.native(p);
  } catch {
    return path.resolve(p);
  }
}

function isOutsideWorkspace(workspaceAbs: string, targetAbs: string): boolean {
  const rel = path.relative(workspaceAbs, targetAbs);
  return rel === ".." || rel.startsWith(`..${path.sep}`);
}

export function analyze(workspace: string, mcpJsonPath: string): PreflightResult {
  const ws = path.resolve(workspace);
  const res: PreflightResult = {
    mcpJsonPath,
    tiedBasePath: "",
    status: Status.OK,
    errors: [],
    warnings: [],
    runtimeNotice: RuntimeMCPNotice,
  };

  let data: string;
  try {
    data = fs.readFileSync(mcpJsonPath, "utf8");
  } catch (err) {
    res.status = Status.Blocked;
    res.errors.push(`read mcp.json: ${String(err)}`);
    return res;
  }

  let root: McpFile;
  try {
    root = JSON.parse(data) as McpFile;
  } catch (err) {
    res.status = Status.Blocked;
    res.errors.push(`parse mcp.json: ${String(err)}`);
    return res;
  }

  if (!root.mcpServers || typeof root.mcpServers !== "object") {
    res.status = Status.Blocked;
    res.errors.push("mcp.json has no mcpServers object");
    return res;
  }

  const raw = root.mcpServers["tied-yaml"];
  if (raw === undefined || raw === null) {
    res.status = Status.Blocked;
    res.errors.push(
      'mcpServers["tied-yaml"] is missing; add it in .cursor/mcp.json (copy_files.sh installs .cursor/skills/tied-yaml only; it does not create mcp.json)',
    );
    return res;
  }

  let srv: TiedYamlServer;
  try {
    srv = raw as TiedYamlServer;
    if (typeof srv !== "object") {
      throw new Error("not an object");
    }
  } catch (err) {
    res.status = Status.Blocked;
    res.errors.push(`parse tied-yaml server entry: ${String(err)}`);
    return res;
  }

  const base = (srv.env?.TIED_BASE_PATH ?? "").trim();
  if (base === "") {
    res.status = Status.Blocked;
    res.errors.push("tied-yaml env.TIED_BASE_PATH is missing or empty");
    return res;
  }

  res.tiedBasePath = base;

  if (!path.isAbsolute(base)) {
    res.status = Status.Blocked;
    res.errors.push(
      `TIED_BASE_PATH must be absolute (got ${JSON.stringify(base)}); see TIED docs for .cursor/mcp.json`,
    );
    return res;
  }

  const wAbs = resolveSymlinkSafe(ws);
  let tAbs: string;
  try {
    tAbs = resolveSymlinkSafe(path.resolve(base));
  } catch (err) {
    res.status = Status.Blocked;
    res.errors.push(`TIED_BASE_PATH does not resolve: ${String(err)}`);
    return res;
  }

  if (isOutsideWorkspace(wAbs, tAbs)) {
    res.status = Status.Blocked;
    res.errors.push(
      `TIED_BASE_PATH (${tAbs}) is outside --workspace (${wAbs}); MCP writes may target the wrong repo`,
    );
    return res;
  }

  const req = path.join(tAbs, "requirements.yaml");
  try {
    const st = fs.statSync(req);
    if (st.isDirectory()) {
      throw new Error("is directory");
    }
  } catch {
    res.warnings.push(
      "no readable tied/requirements.yaml at TIED_BASE_PATH (greenfield or partial bootstrap)",
    );
    if (res.status === Status.OK) {
      res.status = Status.Warning;
    }
  }

  if (res.status === Status.OK && res.warnings.length > 0) {
    res.status = Status.Warning;
  }

  return res;
}

export function run(
  workspace: string,
  mcpJsonOverride: string,
): PreflightResult {
  const mcpPath = locateMcpJson(workspace, mcpJsonOverride);
  return analyze(workspace, mcpPath);
}
